### PR TITLE
Bootstrap profiling indices at startup

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we bootstrap all indices and related structures that are required for Elastic Universal Profiling.

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
This is in preparation of setting up a K/V index scheme that allows the profiling solution to use the out of the box implementation of ILM in Elasticsearch (related to https://github.com/elastic/elasticsearch/issues/85273). What's missing (tackled in follow-up commits in this PR or follow-up PRs)?

* Detection whether a certain structure exists and avoid creating it again (will be tackled in this PR)
* Support for versioning by encoding a version in the index name (possibly in a follow-up PR)
* Implementation of an index scheme for K/V indices that allows to use an ILM policy for these indices as well (follow-up PR)

**Comment by danielmitterdorfer** [comment]
I am a bit uncertain which team would be most appropriate to look at this. I drew inspiration from https://github.com/elastic/elasticsearch/pull/57629 when applying labels. Please feel free to relabel if required.

**Comment by danielmitterdorfer** [comment]
@elasticmachine merge upstream

**Comment by danielmitterdorfer** [comment]
There was an unrelated test failure. I raised https://github.com/elastic/elasticsearch/issues/96000.

**Comment by danielmitterdorfer** [comment]
I'm gonna look separately into `TransformUsageIT#testUsage()` but despite that the PR is ready for review now.

**Comment by danielmitterdorfer** [comment]
I've pushed a preliminary fix for the test failure here (to unblock progress on the PR) and have raised https://github.com/elastic/elasticsearch/pull/96015 separately. I'll incorporate https://github.com/elastic/elasticsearch/pull/96015 once it's merged to `main`.

File: x-pack/plugin/profiler/src/internalClusterTest/java/org/elasticsearch/xpack/profiler/ProfilingTestCase.java
Comment (gmarouli) [inline x-pack/plugin/profiler/src/internalClusterTest/java/org/elasticsearch/xpack/profiler/ProfilingTestCase.java:90]:
```suggestion
            // simple check whether all the indices is present
```

If I am not mistaken this is checking if all the indices have been created, right?

File: x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/profiler/ilm-policy/profiling.json
Comment (danielmitterdorfer) [reply x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/profiler/ilm-policy/profiling.json]: Good point. I've added this in https://github.com/elastic/elasticsearch/pull/95666/commits/7dc6f1ca856384d427785431685813a769ab0c8b.

File: x-pack/plugin/profiler/src/internalClusterTest/java/org/elasticsearch/xpack/profiler/ProfilingTestCase.java
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/internalClusterTest/java/org/elasticsearch/xpack/profiler/ProfilingTestCase.java]: Addressed in https://github.com/elastic/elasticsearch/pull/95666/commits/7dc6f1ca856384d427785431685813a769ab0c8b (I've removed the comment entirely as I think the code is clear enough as it is done now).

File: x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/profiler/ilm-policy/profiling.json
Comment (danielmitterdorfer) [reply x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/profiler/ilm-policy/profiling.json:2]: In this PR I've basically only ported the existing policy as it is defined in https://github.com/elastic/kibana/blob/main/x-pack/plugins/profiling/server/lib/setup/steps/ilm_profiling.json but - as discussed offline - I agree that not everything in there makes sense. I'll address this later in a separate PR.

File: x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/profiler/ilm-policy/profiling.json
Comment (danielmitterdorfer) [reply x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/profiler/ilm-policy/profiling.json]: As `true` is the default anyway, I've removed this now in https://github.com/elastic/elasticsearch/commit/7dc6f1ca856384d427785431685813a769ab0c8b.

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java
Comment (gmarouli) [inline x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java:161]: I am trying to think whether there are other cases that could throw an exception and we need to release the flag. I haven't found one yet but I put it here in case it triggers you or someone else.

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexTemplateRegistry.java
Comment (gmarouli) [inline x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexTemplateRegistry.java:211]: Nit: would you renaming to `areAllTemplatesCreated`?

**Review by gmarouli — APPROVED** [review]
LGTM, @danielmitterdorfer thank you for breaking it into steps :). A few minor comments, that's all

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java]: Good point. I "stole" the original implementation actually from similar code in the base class but yours is definitely more readable. I've changed this now in https://github.com/elastic/elasticsearch/pull/95666/commits/f1dee21951b7bae18cd687618df69e5cd7987cd1.

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java:159]: I did notice that I missed adding a `return` here and added it in https://github.com/elastic/elasticsearch/pull/95666/commits/f1dee21951b7bae18cd687618df69e5cd7987cd1.

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexManager.java]: Done in https://github.com/elastic/elasticsearch/pull/95666/commits/f1dee21951b7bae18cd687618df69e5cd7987cd1

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexTemplateRegistry.java
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingIndexTemplateRegistry.java]: Done in https://github.com/elastic/elasticsearch/pull/95666/commits/f1dee21951b7bae18cd687618df69e5cd7987cd1.

**Comment by danielmitterdorfer** [comment]
Thank you very much for the review @gmarouli!

**Pull request merged**
